### PR TITLE
use early returns in some dtemplate.d methods

### PR DESCRIPTION
**Pull request by thewilsonator**

### CODE REVIEW

File: compiler/src/dmd/dtemplate.d [hunk 1]
```diff
@@ -3843,62 +3843,63 @@ MATCH deduceType(RootObject o, Scope* sc, Type tparam, TemplateParameters* param
...
-                    return;
-                }
             }
+            if (tp && tp.matchArg(sc, t.dim, i, parameters, dedtypes, null) || edim && edim.toInteger() == t.dim.toInteger())
+            {
+                result = deduceType(t.next, sc, tparam.nextOf(), parameters, dedtypes, wm);
+                return;
+            }
+
             visit(cast(Type)t);
```
Comment (PetarKirov) [inline compiler/src/dmd/dtemplate.d:3903]: You moved the call to this function to the `if (!tparam)` case, so shouldn't this be deleted here?
```suggestion
```

File: compiler/src/dmd/dtemplate.d [hunk 2]
```diff
@@ -3923,131 +3924,132 @@ MATCH deduceType(RootObject o, Scope* sc, Type tparam, TemplateParameters* param
...
-
-                    if (!a.isCovariant(t.isref, ap) ||
-                        !deduceType(a.type, sc, ap.type, parameters, dedtypes))
-                    {
-                        result = MATCH.nomatch;
-                        return;
-                    }
-                }
             }
             visit(cast(Type)t);
```
Comment (PetarKirov) [inline compiler/src/dmd/dtemplate.d:4054]: ditto
```suggestion
```

**Comment by RazvanN7** [comment]
ping @thewilsonator

**Comment by dkorpel** [comment]
ping @thewilsonator

**Comment by RazvanN7** [comment]
Maybe such changes should be done in smaller batches such that it is easier to identify the culprit when the test suite fails.

**Pull request closed without merging**